Add MetPy

dopplershift opened the pull request:
I'm not sure if this is the correct procedure, but I'd like to give conda-forge a shot as the channel for hosting MetPy's conda packages.

I've added the recipe for building MetPy, as well as Pint (the only dependency which isn't in conda main).

I'm happy to go through other hoops if there's a better way to go about this. If all goes well, I'd like to get siphon and its larger set of dependencies on here as well.

ocefpaf commented:
> I'm not sure if this is the correct procedure,

Don't worry.  Neither do we :wink: 

> but I'd like to give conda-forge a shot as the channel for hosting MetPy's conda packages.

Great!  Soon the recipes submitted here will become feedstocks (stand alone recipes with their own repository) automatically. You would also be tagged as the recipe maintainer. Nothing prevents others from contributing though.

> I've added the recipe for building MetPy, as well as Pint (the only dependency which isn't in conda main).

Nice.

> I'm happy to go through other hoops if there's a better way to go about this. If all goes well, I'd like to get siphon and its larger set of dependencies on here as well.

Awesome!

Note that @pelson is working hard on this lately. Expect some failures for a while...  I just saw that he changed the CIs from `obvious-ci` to `conda-build-all` and that has nothing to do with your PR.

ocefpaf commented on recipes/metpy/bld.bat:
```diff
@@ -0,0 +1,2 @@
+"%PYTHON%" setup.py install
```
MetPy uses setuptools, right?  You can add ` --single-version-externally-managed --record record.txt` to that command then. (And the one below too.)

dopplershift replied:
Done. Those were in my original recipe, but I couldn't remember the purpose, so I pulled them while "cleaning up". **sigh**

ocefpaf commented on recipes/metpy/meta.yaml:
```diff
@@ -0,0 +1,37 @@
+  fn: metpy-0.3.0.tar.gz
+  url: https://pypi.python.org/packages/source/M/MetPy/MetPy-0.3.0.tar.gz
+  md5: efa24aea506b2d0a829035aea94e6d89
+
```
Add a build field like:

```
build:
    script: python setup.py install --single-version-externally-managed --record record.txt
    number: 0
```

So we can track the build number. (The build script will soon replace `bld.bat` and `build.sh`.)

dopplershift replied:
Done.

pelson commented on recipes/metpy/meta.yaml:
```diff
@@ -0,0 +1,41 @@
+package:
+  name: metpy
+  version: !!str 0.3.0
```
I don't know if this is necessary. @ocefpaf, do you typically just ```"0.3.0"``` it? Or even, just ```0.3.0``` directly.
Either way, one simplification for maintenance of this recipe would be to set a variable up top: https://github.com/conda-forge/staged-recipes/blob/master/ready_recipes/example/meta.yaml#L1

ocefpaf replied:
>  I don't know if this is necessary. @ocefpaf, do you typically just "0.3.0" it? Or even, just 0.3.0 directly.

Right, no need for that.

@dopplershift probably used `conda skeleton` as a template.  (Thee skeleton still creates the yaml file using `!! str` for version.)

> Either way, one simplification for maintenance of this recipe would be to set a variable up top

I don't like the jinja variable when the variable will be used only once like in the example.  Although, in the MetPy recipe, that variable will be used in 3 different places. (`url`, `fn`, and `version`.)

pelson replied:
> I don't like the jinja variable when the variable will be used only once

I agree. Though it is hard to come up with a useful example which doesn't have real source data! :smile:

pelson commented on recipes/metpy/meta.yaml:
```diff
@@ -0,0 +1,41 @@
+    - python
+    - setuptools
+
+# Specify numpy with >= to get it to name packages without numpy dependency
```
That has been fixed for a few versions of conda-build. If you have no compiled dependency, just specifying ```numpy``` will be fine

pelson commented on recipes/metpy/meta.yaml:
```diff
@@ -0,0 +1,41 @@
+
+extra:
+  recipe-maintainers:
+    - dopplershift
```
:+1: :+1: :+1:

pelson commented on recipes/pint/bld.bat:
```diff
@@ -0,0 +1,2 @@
+"%PYTHON%" setup.py install
```
With setuptools, it is pretty much always ```--single-version-externally-manged --record=record.txt```, otherwise we get versioned (eggs?) directories in site-packages.

pelson commented on recipes/metpy/build.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+$PYTHON setup.py install --single-version-externally-managed --record record.txt
```
This file can now be removed.

pelson commented:
> I'm not sure if this is the correct procedure, but I'd like to give conda-forge a shot as the channel for hosting MetPy's conda packages.

Yep, you did good! :smile: 
You also caught the repo in a little bit of upheaval. I've *just* finished opening the floodgates for feedstocks, so now all that is needed is to submit your recipes, get them merged, and be the proud maintainer of a ```conda-forge/metpy-feedstock``` repo. Since you've caught us unawares, for now, feel free to submit the recipes to the ```ready_recipes``` directory (that wont be the case in the future, even 7 days with any luck). From there it will get built, and once merged, converted into a feedstock. I've just updated the front page of https://conda-forge.github.io with a little bit of this information.

dopplershift commented:
Well an accidental rebase required a force push for some reason. But I removed all `build.sh` files and moved to the jinja variables for both metpy and pint (since they both had 3 copies of the version number). Also fixed the setuptools stuff for pint.

pelson commented:
```
BUILD START: pint-0.6-py27_0
Fetching package metadata: ........

WARNING: conda-build appears to be out of date. You have version 1.18.2 but the
latest version is 1.19.0. Run

conda update -n root conda-build

to get the latest version.

Solving package specifications: .............
The following NEW packages will be INSTALLED:

    openssl:    1.0.2f-0      defaults
    pip:        8.0.2-py27_0  defaults
    python:     2.7.11-0      defaults
    readline:   6.2-2         defaults
    setuptools: 19.6.2-py27_0 defaults
    sqlite:     3.9.2-0       defaults
    tk:         8.5.18-0      defaults
    wheel:      0.29.0-py27_0 defaults
    zlib:       1.2.8-0       defaults

Removing old work directory
Source cache directory is: /opt/conda/conda-bld/src_cache
Downloading source to cache: Pint-0.6.zip
Downloading https://pypi.python.org/packages/source/P/Pint/Pint-0.6.zip
Success
Extracting download
Package: pint-0.6-py27_0
source tree in: /opt/conda/conda-bld/work/Pint-0.6
number of files: 0
Fixing permissions
Fixing permissions
BUILD END: pint-0.6-py27_0
TEST START: pint-0.6-py27_0
Fetching package metadata: ........

WARNING: conda-build appears to be out of date. You have version 1.18.2 but the
latest version is 1.19.0. Run

conda update -n root conda-build

to get the latest version.

Solving package specifications: .............
The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    pint-0.6                   |           py27_0           1 KB  file:///opt/conda/conda-bld/linux-64/

The following NEW packages will be INSTALLED:

    openssl:    1.0.2f-0      defaults                             
    pint:       0.6-py27_0    file:///opt/conda/conda-bld/linux-64/
    pip:        8.0.2-py27_0  defaults                             
    python:     2.7.11-0      defaults                             
    readline:   6.2-2         defaults                             
    setuptools: 19.6.2-py27_0 defaults                             
    sqlite:     3.9.2-0       defaults                             
    tk:         8.5.18-0      defaults                             
    wheel:      0.29.0-py27_0 defaults                             
    zlib:       1.2.8-0       defaults                             

===== testing package: pint-0.6-py27_0 =====
import: 'pint'
Traceback (most recent call last):
  File "/opt/conda/conda-bld/test-tmp_dir/run_test.py", line 25, in <module>
    import pint
ImportError: No module named pint
```

pelson commented on ready_recipes/pint/meta.yaml:
```diff
@@ -0,0 +1,32 @@
+  fn: Pint-{{version}}.zip
+  url: https://pypi.python.org/packages/source/P/Pint/Pint-{{version}}.zip
+  md5: 88aa0918019071546b10c5fbdc91ae5e
+
```
Looks like you're missing the:

```
build:
    script: python setup.py install --single-version-externally-managed --record record.txt
    number: 0
```

pelson commented:
It's in there somewhere, but the travis logs tell me:

```
/Users/travis/build/conda-forge/staged-recipes/ready_recipes/metpy is in fine form
/Users/travis/build/conda-forge/staged-recipes/ready_recipes/pint is in fine form
```

dopplershift commented:
Looks like the Travis "create-feedstocks-on-merge" job failed due to a 503--maybe just a temporary/spurious failure?

pelson commented:
Hmmm. I've never seen that one before. The error message is particularly concerning, though it doesn't look like any BINSTAR_TOKEN I've ever seen, so maybe we haven't leaked on that. I'll try restarting.

dopplershift commented:
Same problem as yesterday on win: `no module example`.

pelson commented:
> Same problem as yesterday on win: no module example.

Thanks @dopplershift. I'll get on and fix that. Not something to hold this PR up over.

pelson commented:
> Not something to hold this PR up over.

Ah, except we haven't tested the packaged you've added work. Hmmm. I'll take a look at a fix now.

pelson commented:
Ok, I've now fixed the recipe (in #58). If you rebase we should get the green light (assuming everything builds fine on Windows). Thanks for the patience @dopplershift - this stuff will soon iron out.

dopplershift commented:
Sounds good (I didn't need #59 ? ). No worries about rough edges--the gain from this is **huge** so I'm very willing to be a guinea pig. :smile:

pelson commented:
>  (I didn't need #59 ? )

No. #58 was the workaround. #59 will be the proper fix. Fingers crossed for this one.

dopplershift commented:
Interesting, updating my PR got later win builds to succeed (after the first one failed for the typo.)

dopplershift commented:
Success!

pelson commented:
Thanks @dopplershift. Now we sit back and watch the feedstock creation. :+1:

dopplershift commented:
Really pleased overall with how this process went. I want to mention a couple of issues I've run into:

- Because `metpy` and `pint` went in under the same PR, `metpy`'s package generation failed because `pint` packages weren't (yet) available from anaconda.org. Might be best to dictate future chains of packages go in 1 per PR. Just a thought. Re-running the build for metpy seems to fix it, at least for CircleCI/Travis....
-  Travis wouldn't let me rebuild metpy until I manually had it re-sync with my GitHub account to see that I'm now a part of conda-forge. Might be good to document that somewhere.
- AppVeyor is (of course) a mess. I was able to watch the builds of this repo (staged-recipe) before no problem. But for both pint and metpy, I get "Project not found or access denied." from AppVeyor. (link for MetPy: https://ci.appveyor.com/project/conda-forge/metpy-feedstock/branch/master) Pint at least has packages, so I sense it's a permission issue.

So if I could just kick off windows builds of MetPy, I'd be good to go!

ocefpaf commented:
@dopplershift the AppVeyor access is a pain.  I saw that they have a concept of teams, but I never got around making that work. I will take a second look at that and report back here.

pelson commented:
@dopplershift - that is super useful feedback. Thanks! Worryingly, I can't access the "conda-forge" appveyor metpy either - it seems to have gone under "pelson". (https://ci.appveyor.com/project/pelson/metpy-feedstock/branch/master).
I've re-triggered it (which you could do with a commit I suppose), but I'll take a look at what is going on at the registration phase.

The chaining is an interesting one. Not sure what we can do about it at this point - maybe the feedstock creation needs to hold back on recipes which depend on other recipes which aren't yet feedstocks. Adds a bit of complexity, but is do-able.

> Travis wouldn't let me rebuild metpy until I manually had it re-sync with my GitHub account to see that I'm now a part of conda-forge. Might be good to document that somewhere.

Yes, good idea. Perhaps we could do with a FAQ...

bheklilr commented:
Sorry to drag up an old issue, but I'm not sure where else to ask.  I'm interested in using pint for one of my projects, but I noticed that on the conda-forge channel there are only builds for win-32.  What would be the correct process for getting other platforms built and uploaded?  It looks like the pypi package is flagged for all 3 major platforms and there's a built-in test suite, so I'd imagine that it should be simple to get set up.  Is this something I could do easily myself?  I'm not sure where to start since I've only recently started using conda-forge packages myself.

Thanks!

ocefpaf commented:
> Sorry to drag up an old issue, but I'm not sure where else to ask.

No problem. However, if you want to participate in `pint` check its own feedstock at:

https://github.com/conda-forge/pint-feedstock

> but I noticed that on the conda-forge channel there are only builds for win-32. 

Not sure that is true. See:

http://anaconda.org/conda-forge/pint/files

> I'm not sure where to start since I've only recently started using conda-forge packages myself.

Let me know if we can help you get those packages installed.

bheklilr commented:
@ocefpaf  Odd, for some reason when I looked at it earlier there were only win-32 files, and when I had synced conda-forge to my team's internal anaconda server it only grabbed the win-32 files.  Through the link you posted it definitely is showing all the files now.  I guess it's just a problem on my end, thanks for your help!

ocefpaf commented:
No problem. Glad it worked!
